src/java.base/share/classes/jdk/internal/lang/StableValue.java

```diff
@@ -0,0 +1,429 @@
+
+    /**
+     * If no value is set, sets the stable value to the provided (nullable)
+     * {@code value}, returning the (pre-existing or newly set) value.
+     * <p>
+     * If several threads invoke this method simultaneously, only one thread will succeed
+     * in setting a value and that (witness) value will be returned to all threads.
+     *
+     * @param value to set (nullable)
+     * @return the bound value
+     * @throws NoSuchElementException if {@linkplain #isError()} is {@code true}
+     */
+    V setIfUnset(V value);
+
+    /**
+     * If no value {@linkplain #isSet() is set}, attempts to compute and set a
+     * new (nullable) value using the provided {@code supplier}, returning the
+     * (pre-existing or newly set) value.
+     *
+     * <p>
+     * If the supplier throws an (unchecked) exception, the exception is rethrown, and no
+     * value is set and {@linkplain #isError()} will return {@code true}. The most common
+     * usage is to construct a new object serving as a lazily computed value or memoized
+     * result, as in:
+     *
+     * <pre> {@code
+     * Value witness = stable.computeIfUnset(Value::new);
+     * }</pre>
+     *
+     * @implSpec The implementation logic is equivalent to the following steps for this
+     * {@code stable}:
+     *
+     * <pre> {@code
+     * if (stable.isSet()) {
+     *     return stable.get();
+     * } else {
+     *     V newValue = supplier.get();
+     *     stable.setOrThrow(newValue);
+     *     return newValue;
+     * }
+     * }</pre>
+     * Except it is atomic, thread-safe and will only return the same witness value
+     * regardless if invoked by several threads. Also, the provided {@code supplier}
+     * will only be invoked once even if invoked from several threads and if the
+     * {@code supplier} throws an exception, the stable value will be in an error state.
+     *
+     * // Todo: Should we wrap supplier exceptions into a specific exception type?
+     *
+     * @param  supplier to be used for computing a value
+     * @return the current (pre-existing or computed) value
+     * @throws NoSuchElementException if {@linkplain #isError()} is {@code true}
+     * @throws StackOverflowError if the provided {@code supplier} recursively
+     *         invokes this method upon being invoked.
+     */
+    V computeIfUnset(Supplier<? extends V> supplier);
+
+    // Factories
+
+    /**
+     * {@return a fresh stable value with an unset value}
+     *
+     * @param <V> the value type to set
+     */
+    static <V> StableValue<V> of() {
+        return StableValueImpl.of();
+    }
+
+    /**
+     * {@return a fresh stable value with an unset value where the returned stable's
+     * value is computed in a separate background thread (created via the provided
+     * {@code threadFactory}) using the provided {@code supplier}}
+     * <p>
+     * If the supplier throws an (unchecked) exception, the exception is ignored, and no
+     * value is set.
```

Comment:
Is it likely that users will want to be made aware of failures? If so, perhaps it would make sense to make sure that the Exception hits the UncaughtExceptionHandler? 🤔

Reply:
Good suggestion. An alternative would be to provide an exception listener getting invoked upon hitting an exception.